File: Source/Core/Modes/SearchMode.cs
using System.Linq;

namespace Quip {
  class SearchMode : Mode {
    public SearchMode (DocumentView view)
      : base("/", view) {
      AddMapping(new Keystroke(Key.Escape), LeaveMode);
      AddMapping(new Keystroke(Key.Enter), LeaveMode);
    }

    protected override bool OnHandleKey (Key input, DocumentView view) {
      if (input != Key.Backspace) {
        m_pattern += input.ToText();
      } else {
        m_pattern = m_pattern.Substring(0, m_pattern.Length - 1);
      }

      var selections = view.Document.Search(m_pattern);
      if (selections.Any()) {
        view.Selections.ReplaceWith(selections);
        Name = string.Format("/{0}", m_pattern);
      } else {
        Name = string.Format("/{0} [NOT FOUND]", m_pattern);
      }

      return true;
    }

    bool LeaveMode (DocumentView view) {
      view.Mode = new NormalMode(view);
      return true;
    }

    string m_pattern = string.Empty;
  }
}

File: Source/SearchMode.cpp
#include "SearchMode.hpp"

#include "Document.hpp"
#include "EditContext.hpp"
#include "KeyStroke.hpp"

namespace quip {
  SearchMode::SearchMode () {
    addMapping(Key::Escape, &SearchMode::commitSearch);
    addMapping(Key::Return, &SearchMode::commitSearch);
  }
  
  std::string SearchMode::status () const {
    return "s/" + m_search;
  }
  
  bool SearchMode::onUnmappedKey (const KeyStroke & keyStroke, EditContext & context) {
    if (keyStroke.text().size() > 0) {
      m_search += keyStroke.text();
      
      SelectionSet matches = context.document().matches(m_search);
      context.selections().replace(matches);
      
      return true;
    }
    
    return false;
  }
  
  void SearchMode::commitSearch (EditContext & context) {
    m_search = "";
    
    context.leaveMode();
  }
}